flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/RegularizationType.scala

```diff
@@ -0,0 +1,171 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.optimization
+
+import org.apache.flink.api.scala._
+import org.apache.flink.ml.math.{Vector => FlinkVector, BLAS}
+import org.apache.flink.ml.math.Breeze._
+
+import breeze.numerics._
+import breeze.linalg.{norm => BreezeNorm, max => BreezeMax}
+
+
+
+// TODO(tvas): Change name to RegularizationPenalty?
+/** Represents a type of regularization penalty
+  *
+  * Regularization penalties are used to restrict the optimization problem to solutions with
+  * certain desirable characteristics, such as sparsity for the L1 penalty, or penalizing large
+  * weights for the L2 penalty.
+  *
+  * The regularization term, $R(w)$ is added to the objective function, $f(w) = L(w) + \lambda R(w)$
+  * where $\lambda$ is the regularization parameter used to tune the amount of regularization
+  * applied.
+  */
+abstract class RegularizationType extends Serializable {
+
+  /** Updates the weights by taking a step according to the gradient and regularization applied
+    *
+    * @param oldWeights The weights to be updated
+    * @param gradient The gradient according to which we will update the weights
+    * @param effectiveStepSize The effective step size for this iteration
+    * @param regParameter The regularization parameter to be applied in the case of L1
+    *                     regularization
+    */
+  def takeStep(
+      oldWeights: FlinkVector,
+      gradient: FlinkVector,
+      effectiveStepSize: Double,
+      regParameter: Double) {
+    BLAS.axpy(-effectiveStepSize, gradient, oldWeights)
+  }
+
+  /** Adds regularization to the loss value **/
+  def regLoss(oldLoss: Double, weightVector: FlinkVector, regularizationParameter: Double): Double
+
+}
+
+/** Abstract class for regularization penalties that are differentiable
+  *
+  */
+abstract class DiffRegularizationType extends RegularizationType {
+
+  /** Compute the regularized gradient loss for the given data.
+    * The provided cumGradient is updated in place.
+    *
+    * @param weightVector The current weight vector
+    * @param lossGradient The vector to which the gradient will be added to, in place.
+    * @return The regularized loss. The gradient is updated in place.
+    */
+  def regularizedLossAndGradient(
+      loss: Double,
+      weightVector: FlinkVector,
+      lossGradient: FlinkVector,
+      regularizationParameter: Double) : Double ={
+    val adjustedLoss = regLoss(loss, weightVector, regularizationParameter)
+    regGradient(weightVector, lossGradient, regularizationParameter)
+
+    adjustedLoss
+  }
+
+  /** Adds regularization gradient to the loss gradient. The gradient is updated in place **/
```

Comment:
"Adds the regularization..."

Maybe add docs for parameters and return value. This will be inherited by all implementing subclasses.